plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -229,28 +271,83 @@ private static boolean isLocationPresent(Storage storage)
         return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
     }
 
-    /**
-     * The class enforces intentionally object identity semantics for the value holder,
-     * not value-based class semantics to correctly act as an invalidation guard in the
-     * cache.
-     */
     private static class ValueHolder
     {
-        private final Optional<List<LocatedFileStatus>> files;
+        private final List<LocatedFileStatus> files = synchronizedList(new ArrayList<>());
```

Comment:
I personally find hard to follow the purpose of `ValueHolder` class.
Before this change it was a dumb holder.
Now it acts like a more complex entity : iterator, holder, exception handler...

Can you please consider do some changes to have the logic of the class somehow easier to read?

Reply:
> Now it acts like a more complex entity : iterator, holder, exception handler...

It's still rather simple. It only contains
* files that have already been enumerated
* iterator to enumerate remaining files
* exception if enumeration failed

This is required when we want to cache partially enumerated directories (e.g. same table scanned within single query).
I'm really not sure how this can be simplified significantly. Maybe if we enumerated all files at once, but that would introduce latency issues.

Reply:
> I'd like to understand more about this reasoning.

you run self join query `table_x join table_x`

Both scans might start at same time.
We want to avoid situation where `table_x` directory is listed twice because `DirectoryLister#list` was called twice before full content has been cached